h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -276,33 +331,55 @@ private void validateParams() {
 
         if ("seed".equals(key) || "_seed".equals(key)) continue;  // initialized to the wall clock
 
+        validateParamVals(key);
         // Ugh.  Java callers, like the JUnits or Sparkling Water users, use a leading _.  REST users don't.
-        String prefix = (key.startsWith("_") ? "" : "_");
-
-        // Throw if params has a non-default value which is not in the hyperParams map
-        Object defaultVal = PojoUtils.getFieldValue(_defaultParams, prefix + key, PojoUtils.FieldNaming.CONSISTENT);
-        Object actualVal = PojoUtils.getFieldValue(_params, prefix + key, PojoUtils.FieldNaming.CONSISTENT);
-
-        if (defaultVal != null && actualVal != null) {
-          // both are not set to null
-          if (defaultVal.getClass().isArray() &&
-                  // array
-                  !PojoUtils.arraysEquals(defaultVal, actualVal)) {
-            throw new H2OIllegalArgumentException("Grid search model parameter '" + key + "' is set in both the model parameters and in the hyperparameters map.  This is ambiguous; set it in one place or the other, not both.");
-          } // array
-          if (!defaultVal.getClass().isArray() &&
-                  // ! array
-                  !defaultVal.equals(actualVal)) {
-            throw new H2OIllegalArgumentException("Grid search model parameter '" + key + "' is set in both the model parameters and in the hyperparameters map.  This is ambiguous; set it in one place or the other, not both.");
-          } // ! array
-        } // both are set: defaultVal != null && actualVal != null
-
-        // defaultVal is null but actualVal is not, raise exception
-        if (defaultVal == null && !(actualVal == null)) {
-          // only actual is set
-          throw new H2OIllegalArgumentException("Grid search model parameter '" + key + "' is set in both the model parameters and in the hyperparameters map.  This is ambiguous; set it in one place or the other, not both.");
-        }
       } // for all keys
+      for (Map<String, Object> constraint : _hyperParamConstraints) {
+        for (String key : constraint.keySet()) {
+          // Throw if the user passed an empty value list:
+          Object values = constraint.get(key);
+          if (0 == ((ArrayList) values).toArray().length)
+            throw new H2OIllegalArgumentException("Grid search hyperparameter value list is empty for hyperparameter: " + key);
+
+          if ("seed".equals(key) || "_seed".equals(key)) continue;  // initialized to the wall clock
+
+          if (_hyperParams.get(key) != null) {
+            throw new H2OIllegalArgumentException("Grid search model parameter '" + key + "' is set in " +
+                    "both the constraints and in the hyperparameters map.  This is ambiguous; set it in one place" +
+                    " or the other, not both.");
+          }
+          
+          validateParamVals(key);
+        }
+      }
+    }
```

Comment:
if I'm not mistaken, once you have `_hyperParamConstraints` with the correct type, you can then reuse the existing logic validating `hyperParams` to validate the constraints: 
```
private void validateParams() {
   validateParams(_hyperParams);
   validateParams(_hyperParamsConstraints);
}
```